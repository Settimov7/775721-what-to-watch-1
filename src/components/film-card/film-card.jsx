import * as React from 'react';
import * as PropTypes from 'prop-types';

import {VideoPlayer} from "../video-player/video-player";

const SIZES = {
  width: 280,
  height: 175
};

export class FilmCard extends React.PureComponent {
  constructor(props) {
    super(props);

    this._filmCardRef = React.createRef();

    this.state = {
      timerId: null,
      isPreviewPlaying: false,
    };

    this._mouseEnterHandler = this._mouseEnterHandler.bind(this);
    this._mouseLeaveHandler = this._mouseLeaveHandler.bind(this);
  }

  render() {
    const {film, sizes = SIZES, onTitleClick} = this.props;
    const {name, previewImageSrc, videoSrc} = film;
    const {isPreviewPlaying} = this.state;
    const {width, height} = sizes;

    return (
      <article
        ref={this._filmCardRef}
        className="small-movie-card catalog__movies-card"
        onMouseEnter={this._mouseEnterHandler}
        onMouseLeave={this._mouseLeaveHandler}
      >
        <div className="small-movie-card__image">
          <VideoPlayer
            videoSrc={videoSrc}
            posterSrc={previewImageSrc}
            options={
              {
                width,
                height,
              }
            }
            isPlaying={isPreviewPlaying}
          />
        </div>

        <h3 className="small-movie-card__title">
          <a
            className="small-movie-card__link"
            href="movie-page.html"
            onClick={onTitleClick}
          >
            {name}
          </a>
        </h3>
      </article>
    );
  }

  _mouseEnterHandler() {
    const timerId = setTimeout(() => {
      this.setState({
        isPreviewPlaying: true,
      });
    }, 1000);

    this.setState({
      timerId
    });
  }

  _mouseLeaveHandler() {
    clearTimeout(this.state.timerId);

    this.setState({
      isPreviewPlaying: false,
      timerId: null,
    });
  }
}

FilmCard.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    posterImageSrc: PropTypes.string,
    previewImageSrc: PropTypes.string,
    backgroundImageSrc: PropTypes.string,
    backgroundColor: PropTypes.string,
    description: PropTypes.string,
    rating: PropTypes.number,
    scores: PropTypes.number,
    director: PropTypes.string,
    starring: PropTypes.arrayOf(PropTypes.string),
    runTime: PropTypes.number,
    genre: PropTypes.string,
    releasedYear: PropTypes.number,
    isFavorite: PropTypes.bool,
    videoSrc: PropTypes.string,
    previewVideoSrc: PropTypes.string,
  }),
  sizes: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number,
  }),
  onTitleClick: PropTypes.func,
};
