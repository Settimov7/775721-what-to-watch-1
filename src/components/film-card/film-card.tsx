import * as React from 'react';
import {connect} from 'react-redux';

import {SvgSprite} from '../svg-sprite/svg-sprite';
import {Logo} from '../logo/logo';
import UserBlock from '../user-block/user-block';

import {Operation} from '../../reducer/films/films';

import {Film} from '../../types';

interface Props {
  film: Film;
  onPlayButtonClick: () => void;
}

export const FilmCard: React.FunctionComponent<Props> = (props) => {
  const {film, onPlayButtonClick} = props;
  const {name, backgroundImageSrc, genre, releasedYear, posterImageSrc} = film;

  return (
    <React.Fragment>
      <SvgSprite />

      <section className="movie-card">
        <div className="movie-card__bg">
          <img src={backgroundImageSrc} alt={name}/>
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header movie-card__head">
          <Logo />

          <UserBlock />
        </header>

        <div className="movie-card__wrap">
          <div className="movie-card__info">
            <div className="movie-card__poster">
              <img src={posterImageSrc} alt={name} width="218"
                   height="327"/>
            </div>

            <div className="movie-card__desc">
              <h2 className="movie-card__title">{name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{releasedYear}</span>
              </p>

              <div className="movie-card__buttons">
                <button className="btn btn--play movie-card__button" type="button" onClick={(evt) => {
                  evt.preventDefault();

                  onPlayButtonClick();
                }}>
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"/>
                  </svg>
                  <span>Play</span>
                </button>

                <button className="btn btn--list movie-card__button" type="button">
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"/>
                  </svg>
                  <span>My list</span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </section>
    </React.Fragment>
  );
};

const mapDispatchToProps = (dispatch) => ({
  onAddFilmToFavorites: (filmId): void => {
    dispatch(Operation.changeFilmFavoriteStatus(filmId, true))
  }
});

export default connect(null, mapDispatchToProps)(FilmCard);
