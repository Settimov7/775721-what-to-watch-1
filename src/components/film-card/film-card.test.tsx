import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {FilmCard} from "./film-card";

const mock = {
  film: {
    id: 1,
    name: `Film-1`,
    posterImageSrc: `/image.jpg`,
    previewImageSrc: `/image.jpg`,
    backgroundImageSrc: `/image.jpg`,
    backgroundColor: `#123456`,
    description: `description`,
    rating: 5,
    scores: 123456,
    director: `director`,
    starring: [`person-1`, `person-2`, `person-3`],
    runTime: 99,
    genre: `Genre`,
    releasedYear: 2019,
    isFavorite: false,
    videoSrc: `video.mp4`,
    previewVideoSrc: `video.mp4`,
  },
};

it(`Film card correctly renders`, () => {
  const props = {
    film: mock.film,
  };

  const card = renderer.create(<FilmCard {...props}/>).toJSON();

  expect(card).toMatchSnapshot();
});
