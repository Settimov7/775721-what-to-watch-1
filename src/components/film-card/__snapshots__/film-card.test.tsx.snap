// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film card correctly renders 1`] = `
<article
  className="small-movie-card catalog__movies-card"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="small-movie-card__image"
  >
    <video
      height={175}
      loop={true}
      muted={true}
      poster="/image.jpg"
      src="video.mp4"
      width={280}
    />
  </div>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="movie-page.html"
    >
      Film-1
    </a>
  </h3>
</article>
`;
